README.md

```diff
@@ -1,2 +1,175 @@
+or simply don't support the same transports (e.g. go-ipfs vs. browser-ipfs).
+libp2p already has modules for NAT ([go-libp2p-nat][go-libp2p-nat]),
+but these don't always do the job, just because NAT traversal is a clusterfuck.
+That's why it's useful to have a simple relay protocol.
+
+One node asks a relay node to connect to another node on behalf of the first node.
+The relay node shortcircuits its streams to the two nodes,
+and they are then connected through the relay.
+This connection quaks and behaves just like a direct connection would,
+because it is in fact a direct connection, but not over e.g. a TCP stream,
+but over an existing swarm stream with another peer.
+
+Relayed connections are end-to-end encrypted just like direct connections.
+
+The circuit relay is both a tunneled transport and a mounted swarm protocol.
+The transport is the means to *establishing* and *accepting* connections,
+and the swarm protocol is the means to *relaying* connections.
+
+```
++-------+    /ip4/.../tcp/.../ws/ipfs/QmRelay    +---------+    /ip4/.../tcp/ipfs/QmTwo    +-------+
+| QmOne | <------------------------------------> | QmRelay | <---------------------------> | QmTwo |
++-------+                                        +---------+                               +-------+
+      ^                                            +-----+                                    ^
+      |                                            |     |                                    |
+      |        /libp2p-circuit-relay/QmTwo         |     |                                    |
+      +--------------------------------------------+     +------------------------------------+
+```
+
+[go-libp2p-nat]: https://github.com/libp2p/go-libp2p-nat
+
+
+## Dramatization
+
+Cast:
+- QmOne, the dialing node (browser).
+- QmTwo, the listening node (go-ipfs).
+- QmRelay, a node which speaks the circuit relay protocol (go-ipfs or js-ipfs).
+
+Scene 1:
+- QmOne wants to connect to QmTwo,
+  and through peer routing has acquired a set of addresses of QmTwo.
+- QmTwo doesn't support any of the transports used by QmOne.
+- Awkward silence.
+
+Scene 2:
+- All three nodes have learned to speak the `/ipfs/relay/circuit` protocol.
+- QmRelay is configured to allow relaying connections between other nodes.
+- QmOne is configured to use QmRelay for relaying.
+- QmOne automatically added `/libp2p-circuit-relay/ipfs/QmTwo` to its set of QmTwo addresses.
+- QmOne tries to connect via relaying, because it shares this transport with QmTwo.
+- A lively and prolonged dialogue ensues.
+
+
+### Addressing
+
+Q: should /relay encapsulate /ipfs?
+  - such that /libp2p-circuit-relay/ipfs/Qmfoobar
+  - might be mandated by relay-as-a-transport...
+  - A: yes
+  - need to update the multiaddr size from V to 0
+Q: should the relay transport be configured a a simple listener?
+  - e.g. simply on /libp2p-circuit-relay
+  - or should it just be enabled by default?
```

Comment:
I believe the question/answer is that we need to migrate Swarm addrs to be:
- addrs to listen
- addrs to announce
- addrs to not announce

Relay would be an addr to announce